File: engine/static/render/src/2d/scene_2d.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/color.hpp"
#include "argus/lowlevel/memory.hpp"

#include "argus/render/2d/camera_2d.hpp"
#include "argus/render/common/scene.hpp"
#include "argus/render/2d/render_group_2d.hpp"
#include "argus/render/2d/render_object_2d.hpp"
#include "argus/render/2d/scene_2d.hpp"
#include "internal/render/common/scene.hpp"
#include "internal/render/pimpl/2d/light_2d.hpp"
#include "internal/render/pimpl/2d/render_object_2d.hpp"
#include "internal/render/pimpl/2d/scene_2d.hpp"
#include "internal/render/module_render.hpp"

#include <stdexcept>
#include <string>
#include <vector>
#include <algorithm>

namespace argus {
    // forward declarations
    class Canvas;

    class RenderPrim2D;

    class Transform2D;

    struct pimpl_Scene;

    static PoolAllocator g_pimpl_pool(sizeof(pimpl_Scene2D));

    Scene2D &Scene2D::create(const std::string &id) {
        if (g_scenes.find(id) != g_scenes.end()) {
            throw std::invalid_argument("Scene with given ID already exists");
        }

        auto scene = new Scene2D(id, Transform2D());
        g_scenes.insert({id, scene});

        return *scene;
    }

    Scene2D::Scene2D(const std::string &id, const Transform2D &transform) :
            Scene(SceneType::TwoD),
            m_pimpl(&g_pimpl_pool.construct<pimpl_Scene2D>(id, *this, transform)) {
    }

    Scene2D::Scene2D(Scene2D &&rhs) noexcept:
            Scene(SceneType::TwoD),
            m_pimpl(rhs.m_pimpl) {
        rhs.m_pimpl = nullptr;
    }

    Scene2D::~Scene2D(void) {
        g_scenes.erase(g_scenes.find(m_pimpl->id), g_scenes.end());

        if (m_pimpl != nullptr) {
            g_pimpl_pool.destroy(m_pimpl);
        }
    }

    pimpl_Scene *Scene2D::get_pimpl(void) const {
        return dynamic_cast<pimpl_Scene *>(m_pimpl);
    }

    bool Scene2D::is_lighting_enabled(void) {
        return m_pimpl->lighting_enabled;
    }

    void Scene2D::set_lighting_enabled(bool enabled) {
        m_pimpl->lighting_enabled = enabled;
    }

    float Scene2D::peek_ambient_light_level(void) const {
        return m_pimpl->ambient_light_level.peek();
    }

    ValueAndDirtyFlag<float> Scene2D::get_ambient_light_level(void) {
        return m_pimpl->ambient_light_level.read();
    }

    void Scene2D::set_ambient_light_level(float level) {
        m_pimpl->ambient_light_level = level;
    }

    const Vector3f &Scene2D::peek_ambient_light_color(void) const {
        return m_pimpl->ambient_light_color.peek();
    }

    ValueAndDirtyFlag<Vector3f> Scene2D::get_ambient_light_color(void) {
        return m_pimpl->ambient_light_color.read();
    }

    void Scene2D::set_ambient_light_color(const Vector3f &color) {
        m_pimpl->ambient_light_color = normalize_rgb(color);
    }

    std::vector<std::reference_wrapper<Light2D>> Scene2D::get_lights(void) {
        std::vector<std::reference_wrapper<Light2D>> res;
        res.reserve(m_pimpl->lights_staging->size());
        std::transform(m_pimpl->lights_staging->begin(), m_pimpl->lights_staging->end(), std::back_inserter(res),
                [](auto &kvp) { return std::reference_wrapper(kvp.second); });
        return res;
    }

    std::vector<std::reference_wrapper<const Light2D>> Scene2D::get_lights_for_render(void) {
        std::vector<std::reference_wrapper<const Light2D>> res;
        res.reserve(m_pimpl->lights->size());
        std::transform(m_pimpl->lights->cbegin(), m_pimpl->lights->cend(), std::back_inserter(res),
                [](const auto &kvp) { return std::reference_wrapper(kvp.second); });
        return res;
    }

    Handle Scene2D::add_light(Light2DType type, bool is_occludable, const Vector3f &color,
            LightParameters params, const Transform2D &iniital_transform) {
        Light2D light(type, is_occludable, normalize_rgb(color), params, iniital_transform);
        auto inserted = m_pimpl->lights_staging->insert({ light.get_handle(), light });
        return inserted.first->first;
    }

    std::optional<std::reference_wrapper<Light2D>> Scene2D::get_light(Handle handle) {
        auto *ptr = g_render_handle_table.deref<Light2D>(handle);
        return ptr != nullptr ? std::make_optional(std::reference_wrapper(*ptr)) : std::nullopt;
    }

    void Scene2D::remove_light(Handle handle) {
        m_pimpl->lights_staging->erase(handle);
    }

    std::optional<std::reference_wrapper<RenderGroup2D>> Scene2D::get_group(Handle handle) {
        auto *ptr = g_render_handle_table.deref<RenderGroup2D>(handle);
        return ptr != nullptr ? std::make_optional(std::reference_wrapper(*ptr)) : std::nullopt;
    }

    std::optional<std::reference_wrapper<RenderObject2D>> Scene2D::get_object(Handle handle) {
        auto *ptr = g_render_handle_table.deref<RenderObject2D>(handle);
        return ptr != nullptr ? std::make_optional(std::reference_wrapper(*ptr)) : std::nullopt;
    }

    Handle Scene2D::add_group(const Transform2D &transform) {
        return m_pimpl->root_group_write->add_group(transform);
    }

    Handle Scene2D::add_object(const std::string &material, const std::vector<RenderPrim2D> &primitives,
            const Vector2f &anchor_point, const Vector2f &atlas_stride, uint32_t z_index, float light_opacity,
            const Transform2D &transform) {
        return m_pimpl->root_group_write->add_object(material, primitives, anchor_point,
                atlas_stride, z_index, light_opacity, transform);
    }

    void Scene2D::remove_group(Handle handle) {
        m_pimpl->root_group_write->remove_group(handle);
    }

    void Scene2D::remove_object(Handle handle) {
        m_pimpl->root_group_write->remove_object(handle);
    }

    std::optional<std::reference_wrapper<Camera2D>> Scene2D::find_camera(const std::string &id) const {
        auto it = m_pimpl->cameras.find(id);
        return it != m_pimpl->cameras.cend() ? std::make_optional(std::reference_wrapper(it->second)) : std::nullopt;
    }

    Camera2D &Scene2D::create_camera(const std::string &id) {
        if (m_pimpl->cameras.find(id) != m_pimpl->cameras.end()) {
            throw std::invalid_argument("Camera with provided ID already exists in scene");
        }

        auto it = m_pimpl->cameras.insert({id, Camera2D(id, *this)});

        return it.first->second;
    }

    void Scene2D::destroy_camera(const std::string &id) {
        auto it = m_pimpl->cameras.find(id);
        if (it == m_pimpl->cameras.end()) {
            throw std::invalid_argument("Camera with provided ID does not exist in scene");
        }

        m_pimpl->cameras.erase(it);
    }

    void Scene2D::lock_render_state(void) {
        m_pimpl->read_lock.lock();
    }

    void Scene2D::unlock_render_state(void) {
        m_pimpl->read_lock.unlock();
    }
}

File: engine/auxiliary/render_rustabi/src/argus/render/twod/scene_2d.rs
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
use std::ffi::CStr;
use std::ptr;
use lowlevel_rustabi::argus::lowlevel::{Handle, ValueAndDirtyFlag, Vector2f, Vector3f};
use lowlevel_rustabi::util::str_to_cstring;
use crate::argus::render::*;
use crate::render_cabi::*;

#[derive(Clone)]
pub struct Scene2d {
    handle: argus_scene_2d_t,
}

impl Scene2d {
    pub(crate) fn of(handle: argus_scene_2d_t) -> Self {
        Self { handle }
    }

    pub(crate) fn get_ffi_handle(&self) -> argus_scene_2d_const_t {
        self.handle
    }

    pub fn create(id: &str) -> Self {
        unsafe {
            let id_c = str_to_cstring(id);
            Self::of(argus_scene_2d_create(id_c.as_ptr()))
        }
    }

    pub fn get_id(&self) -> String {
        unsafe { CStr::from_ptr(argus_scene_2d_get_id(self.handle)).to_string_lossy().to_string() }
    }
    
    pub fn is_lighting_enabled(&self) -> bool {
        unsafe { argus_scene_2d_is_lighting_enabled(self.handle) }
    }

    pub fn set_lighting_enabled(&mut self, enabled: bool) {
        unsafe { argus_scene_2d_set_lighting_enabled(self.handle, enabled); }
    }

    pub fn peek_ambient_light_level(&self) -> f32 {
        unsafe { argus_scene_2d_peek_ambient_light_level(self.handle) }
    }

    pub fn get_ambient_light_level(&mut self) -> ValueAndDirtyFlag<f32> {
        unsafe {
            let mut dirty = false;
            let level = argus_scene_2d_get_ambient_light_level(self.handle, &mut dirty);
            ValueAndDirtyFlag::of(level, dirty)
        }
    }

    pub fn set_ambient_light_level(&mut self, level: f32) {
        unsafe { argus_scene_2d_set_ambient_light_level(self.handle, level); }
    }

    pub fn peek_ambient_light_color(&self) -> Vector3f {
        unsafe { argus_scene_2d_peek_ambient_light_color(self.handle).into() }
    }

    pub fn get_ambient_light_color(&mut self) -> ValueAndDirtyFlag<Vector3f> {
        unsafe {
            let mut dirty = false;
            let color = argus_scene_2d_get_ambient_light_color(self.handle, &mut dirty).into();
            ValueAndDirtyFlag::of(color, dirty)
        }
    }

    pub fn set_ambient_light_color(&mut self, color: Vector3f) {
        unsafe { argus_scene_2d_set_ambient_light_color(self.handle, color.into()); }
    }

    pub fn get_lights(&mut self) -> Vec<Light2d> {
        unsafe {
            let count = argus_scene_2d_get_lights_count(self.handle);
            let mut lights: Vec<argus_light_2d_t> = Vec::with_capacity(count);
            lights.resize(count, ptr::null_mut());

            argus_scene_2d_get_lights(self.handle, lights.as_mut_ptr(), count);
            lights.into_iter().map(Light2d::of).collect()
        }
    }

    pub fn get_lights_for_render(&mut self) -> Vec<Light2d> {
        unsafe {
            let count = argus_scene_2d_get_lights_count_for_render(self.handle);
            let mut lights: Vec<argus_light_2d_const_t> = Vec::with_capacity(count);
            lights.resize(count, ptr::null_mut());

            argus_scene_2d_get_lights_for_render(self.handle, lights.as_mut_ptr(), count);
            lights.into_iter().map(|l| Light2d::of(l as argus_light_2d_t)).collect()
        }
    }

    pub fn add_light(
        &mut self,
        light_type: Light2dType,
        is_occludable: bool,
        color: Vector3f,
        params: Light2dParameters,
        initial_transform: Transform2d
    ) -> Handle {
        unsafe {
            argus_scene_2d_add_light(
                self.handle,
                light_type.into(),
                is_occludable,
                color.into(),
                params.into(),
                initial_transform.into()
            ).into()
        }
    }

    pub fn get_light(&mut self, light_handle: Handle) -> Option<Light2d> {
        unsafe {
            argus_scene_2d_get_light(self.handle, light_handle.into())
                .as_mut()
                .map(|ptr| Light2d::of(ptr))
        }
    }

    pub fn remove_light(&mut self, light_handle: Handle) {
        unsafe { argus_scene_2d_remove_light(self.handle, light_handle.into()); }
    }

    pub fn get_group(&mut self, group_handle: Handle) -> Option<RenderGroup2d> {
        unsafe {
            argus_scene_2d_get_group(self.handle, group_handle.into())
                .as_mut()
                .map(|ptr| RenderGroup2d::of(ptr))
        }
    }

    pub fn get_object(&mut self, obj_handle: Handle) -> Option<RenderObject2d> {
        unsafe {
            argus_scene_2d_get_object(self.handle, obj_handle.into())
                .as_mut()
                .map(|ptr| RenderObject2d::of(ptr))
        }
    }

    pub fn add_group(&mut self, transform: ArgusTransform2d) -> Handle {
        unsafe { argus_scene_2d_add_group(self.handle, transform.into()).into() }
    }

    pub fn add_object(
        &mut self,
        material: &str,
        primitives: &Vec<RenderPrimitive2d>,
        anchor_point: Vector2f,
        atlas_stride: Vector2f,
        z_index: u32,
        light_opacity: f32,
        transform: Transform2d
    ) -> Handle {
        unsafe {
            let prims_count = primitives.len();
            let prims: Vec<ArgusRenderPrimitive2d> = primitives.into_iter()
                .map(|prim| prim.into())
                .collect();

            let material_c = str_to_cstring(material);
            
            argus_scene_2d_add_object(
                self.handle,
                material_c.as_ptr(),
                prims.as_ptr(),
                prims_count,
                anchor_point.into(),
                atlas_stride.into(),
                z_index,
                light_opacity,
                transform.into(),
            ).into()
        }
    }

    pub fn remove_group(&mut self, group_handle: Handle) {
        unsafe { argus_scene_2d_remove_group(self.handle, group_handle.into()); }
    }

    pub fn remove_object(&mut self, obj_handle: Handle) {
        unsafe { argus_scene_2d_remove_object(self.handle, obj_handle.into()); }
    }

    pub fn find_camera(&self, id: &str) -> Camera2d {
        unsafe {
            let id_c = str_to_cstring(id);
            Camera2d::of(argus_scene_2d_find_camera(self.handle, id_c.as_ptr()))
        }
    }

    pub fn create_camera(&mut self, id: &str) -> Camera2d {
        unsafe {
            let id_c = str_to_cstring(id);
            Camera2d::of(argus_scene_2d_create_camera(self.handle, id_c.as_ptr()))
        }
    }

    pub fn destroy_camera(&mut self, id: &str) {
        unsafe {
            let id_c = str_to_cstring(id);
            argus_scene_2d_destroy_camera(self.handle, id_c.as_ptr());
        }
    }

    pub fn lock_render_state(&mut self) {
        unsafe { argus_scene_2d_lock_render_state(self.handle); }
    }

    pub fn unlock_render_state(&mut self) {
        unsafe { argus_scene_2d_unlock_render_state(self.handle); }
    }
}

impl Into<Scene> for Scene2d {
    fn into(self) -> Scene {
        Scene::of(self.handle)
    }
}
